packages/slate/src/utils/string.ts

```diff
@@ -5,136 +5,224 @@
     }
 
-    if (isBMPEmoji(charCode)) {
-      if (prev && prev !== 'ZWJ' && prev !== 'VAR') {
+    if (isBMPEmoji(code)) {
+      if (isLTR && prev && prev !== 'ZWJ' && prev !== 'VAR') {
+        break
+      }
+
+      dist += codepoint.length
+
+      if (isRTL && prev === 'VAR') {
         break
       }
-      offset += 1
+
       prev = 'BMP'
-      charCode = text.charCodeAt(offset)
+      continue
+    }
+
+    if (isModifier(code)) {
+      dist += codepoint.length
+      prev = 'MOD'
+
+      continue
+    }
+
+    const [isTagStart, isTagEnd] = isLTR
+      ? [isBlackFlag, isCancelTag]
+      : [isCancelTag, isBlackFlag]
+    if (isTagStart(code)) {
+      if (prev === 'TAG') break
 
+      dist += codepoint.length
+      prev = 'TAG'
+      continue
+    }
+    if (isTagEnd(code)) {
+      dist += codepoint.length
+      break
+    }
+    if (prev === 'TAG' && isTag(code)) {
+      dist += codepoint.length
       continue
     }
 
-    if (isVariationSelector(charCode)) {
-      if (prev && prev !== 'ZWJ') {
+    if (isRegionalIndicator(code)) {
+      dist += codepoint.length
+
+      if (prev === 'RI') {
         break
       }
-      offset += 1
-      prev = 'VAR'
-      charCode = text.charCodeAt(offset)
+
+      prev = 'RI'
+
+      continue
+    }
+
+    if (!isBMP(code)) {
+      // If previous code point is not sequenceable, it means we are not in a
+      // sequence.
+      if (prev === 'NSEQ') {
+        break
+      }
+
+      dist += codepoint.length
+      prev = 'NSEQ'
+
       continue
     }
 
     // Modifier 'groups up' with what ever character is before that (even whitespace), need to
     // look ahead.
-    if (prev === 'MOD') {
-      offset += 1
+    if (isLTR && prev === 'MOD') {
+      dist += codepoint.length
       break
     }
 
     // If while loop ever gets here, we're done (e.g latin chars).
     break
   }
 
-  return offset || 1
+  return dist || 1
 }
 
 /**
  * Get the distance to the end of the first word in a string of text.
  */
 
-export const getWordDistance = (text: string): number => {
-  let length = 0
-  let i = 0
+export const getWordDistance = (text: string, isRTL = false): number => {
+  let dist = 0
   let started = false
-  let char
 
-  while ((char = text.charAt(i))) {
-    const l = getCharacterDistance(char)
-    char = text.slice(i, i + l)
-    const rest = text.slice(i + l)
+  while (text.length > 0) {
+    const charDist = getCharacterDistance(text, isRTL)
+    const [char, remaining] = split(text, charDist, isRTL)
 
-    if (isWordCharacter(char, rest)) {
+    if (isWordCharacter(char, remaining, isRTL)) {
       started = true
-      length += l
+      dist += charDist
     } else if (!started) {
-      length += l
+      dist += charDist
     } else {
       break
     }
 
-    i += l
+    text = remaining
+  }
+
+  return dist
+}
+
+/**
+ * Split a string at a given distance starting from the end when `isRTL` is set
+ * to `true`.
+ */
+
+export const split = (
```

Comment:
I'd prefer a name that's more specific since split is a common JS term. splitByCharacterDistance or something maybe?